kam2doc-lib/src/main/java/it/facile/kam2doc/cropping/CropView.kt

```diff
@@ -0,0 +1,103 @@
+import android.graphics.Bitmap
+import android.util.AttributeSet
+import android.view.Gravity
+import android.view.LayoutInflater
+import android.widget.FrameLayout
+import androidx.core.view.doOnNextLayout
+import it.facile.kam2doc.InvalidPointsException
+import it.facile.kam2doc.R
+import it.facile.kam2doc.Result
+import it.facile.kam2doc.utils.ImageResizer
+import it.facile.kam2doc.utils.ScanUtils
+import it.facile.kam2doc.utils.cropBitmapFromPoints
+import it.facile.kam2doc.utils.getBitmapFromLocation
+import kotlinx.android.synthetic.main.layout_crop.view.*
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+class CropView @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : FrameLayout(context, attrs, defStyleAttr) {
+
+    lateinit var currentImagePath: String
+
+    private val coroutineScope = CoroutineScope(Job())
+    private val currentImage: Bitmap
+        get() = getBitmapFromLocation(currentImagePath)
+    private val polygonPadding: Int
+        get() = polygonView.resources.getDimension(R.dimen.scanPadding).toInt()
+
+    init {
+        LayoutInflater.from(context).inflate(R.layout.layout_crop, this)
+    }
+
+    fun setup(imagePath: String) {
+        currentImagePath = imagePath
+
+        doOnNextLayout {
+            val scaledBitmap =
+                ImageResizer.scaleBitmap(currentImage, sourceFrame.width, sourceFrame.height)
+            sourceImageView.setImageBitmap(scaledBitmap)
+            val layoutParams = LayoutParams(
+                scaledBitmap.width + 2 * polygonPadding,
+                scaledBitmap.height + 2 * polygonPadding
+            )
+            layoutParams.gravity = Gravity.CENTER
+            polygonView.layoutParams = layoutParams
+            coroutineScope.launch {
+                setupInitialPoints(
+                    imageWidth = scaledBitmap.width,
+                    imageHeight = scaledBitmap.height
+                )
+            }
+        }
+    }
+
+    private suspend fun setupInitialPoints(imageWidth: Int, imageHeight: Int) {
+        withContext(Dispatchers.Default) {
+            val (_, points) = cropBitmapFromPoints(currentImage, null)
+            val scaledPoints =
+                ScanUtils.downScalePoints(
+                    points = points,
+                    bitmap = currentImage,
+                    targetImgWidth = imageWidth,
+                    targetImgHeight = imageHeight
+                )
+            withContext(Dispatchers.Main) { polygonView.setPoints(scaledPoints) }
+        }
+    }
+
+    fun crop(): Result<Bitmap, InvalidPointsException> {
+        val originalBitmap = getBitmapFromLocation(currentImagePath)
+        val points = polygonView.getPoints()
+        return if (PolygonUtils.isValidShape(points)) {
```

Comment:
since there are many lines of code, I would use a guard
```suggestion
       if (PolygonUtils.isValidShape(points).not()) {
           return Result.Failure(InvalidPointsException())
       }
```
not a CR